feat: multi-hop step accessor safety

gomesalexandre opened the pull request:
## Description

Follows-up on https://github.com/shapeshift/web/pull/6765 with `getHopByIndex` introducing hop access logic consolidation and safety:

- Adds a new `SupportedTradeQuoteStepIndex` so we make accessors safe by only being to access items at 0th and 1st position, since we don't support multi-hops of more than 2 hops currently). This only ensures easier migration to multi-hops consisting of more than 2 hops when we do support that.
- Adds a new `getHopByIndex` that:
  - narrows the return type by the input `quote` being undefined (or not)
  - leverages the type-safety above to ensure we only access elements that aren't out of bounds
  - ensures we don't try to access an index > 0 for single-hop quotes
 
<!-- Please describe your changes -->

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

- relates to the closed https://github.com/shapeshift/web/issues/6764

## Risk
> High Risk PRs Require 2 approvals

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

WARNING: If your PR introduces a new on-chain transaction or modifies an existing one, please add the "High-Risk" label to this PR and ask for 2 reviewers to approve it before merging.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

> What protocols, transaction types or contract interactions might be affected by this PR?

Very low, this should come as no runtime change, do a small regression of swapper end-to-end as paranoia

## Testing

<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>

- Swapper is still happy for single hops
- Swapper is still happy for Li.Fi multi-hops (see PR under this one i.e https://github.com/shapeshift/web/pull/6765)

### Engineering

<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

- ☝🏽 

### Operations

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

- ☝🏽 

## Screenshots (if applicable)

<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"feat_lifi_multi_hop","parentHead":"12641cee43375267da66ab9ad4f8f9e97b497c92","parentPull":6765,"trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"feat_lifi_multi_hop","parentHead":"82835e7e8b6b9ad304ac56367b8a62cadce85686","parentPull":6765,"trunk":"develop"}
```
-->

0xApotheosis commented on src/state/slices/tradeQuoteSlice/helpers.ts:
```diff
@@ -106,6 +117,56 @@ export const getTotalProtocolFeeByAsset = (quote: TradeQuote): Record<AssetId, P
 }
+
+export const sortTradeQuotes = (
+  tradeQuotes: Partial<Record<SwapperName, Record<string, ApiQuote>>>,
```
```suggestion
  tradeQuotes: PartialRecord<SwapperName, Record<string, ApiQuote>>,
```

gomesalexandre replied:
0f5a515299

0xApotheosis reviewed (changes requested):
I'm currently not able to get quotes on this branch, tried both UTXOs and EVMs, with single and cross-account trades.

<img width="1812" alt="Screenshot 2024-05-02 at 3 18 23 PM" src="https://github.com/shapeshift/web/assets/97164662/f0aa0ecf-fbdf-42ab-b1a6-47ac6527e305">

<img width="1780" alt="Screenshot 2024-05-02 at 3 20 18 PM" src="https://github.com/shapeshift/web/assets/97164662/788cd55f-8622-4c81-92cb-f2c79d10160a">

gomesalexandre commented:
> I'm currently not able to get quotes on this branch, tried both UTXOs and EVMs, with single and cross-account trades.

@0xApotheosis as discussed, way too much "safety" in this PR indeed, voiding very valid accessors at runtime. Fixed in 74ad060060 - and also removed the whole generics dance while at it. The whole `if trade quote is defined, hop is defined` was actually invalid. 

This prompted us to "fix" types in c437ee2da686cdb40220e4ec464f2dd3093eee20 - note `fix` in quotation marks here as most of this is really explicitly checking for undefineds - or adding non-null assertion for first and last hops - which we were not doing before.

Confirmed single hop quotes are happy again: 
<img width="979" alt="image" src="https://github.com/shapeshift/web/assets/17035424/c35e159b-214b-4f64-918f-4cf04ca94362">

And so are multi-hops:

<img width="927" alt="image" src="https://github.com/shapeshift/web/assets/17035424/d7ae9d1f-11d0-4962-b84e-9c11305a4f3c">


Also did a full e2e regression test of a single-hop:

https://github.com/shapeshift/web/assets/17035424/27b9faa3-cd84-4b5a-9c89-94ba1ad15550

And of a multi-hop:

<img width="434" alt="image" src="https://github.com/shapeshift/web/assets/17035424/01e37cc5-599c-4799-8f83-3398ee47cc7a">

0xApotheosis reviewed (approved):
There we go ✅ 

Checked both single and multi-hop trades on both EVM and UTXO assets.
All are now happy.

<img width="345" alt="Screenshot 2024-05-03 at 9 58 05 AM" src="https://github.com/shapeshift/web/assets/97164662/c3da16dd-cb66-4074-8488-e74ff1c5d45e">
